### PR TITLE
Create a python script to deploy Kubeflow on GCP via deployment manager.

### PR DESCRIPTION
**Pull request by jlewi**
* The scripts replaces our bash commands
* For teardown we want to add retries to better handle INTERNAL_ERRORS
  with deployment manager that are causing the test to be flaky.

Related to #836 verify Kubeflow deployed correctly with deployment manager.

* Fix resource_not_found errors in delete (#833)

* The not found error was due to the type providers for K8s resources
  being deleted before the corresponding K8s resources. So the subsequent
  delete of the K8s resource would fail because the type provider did not
  exist.

* We fix this by using a $ref to refer to the type provider in the type field
  of K8s resources.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/866)

### CODE REVIEW

File: testing/deploy_kubeflow_gcp.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+def main():
+  test_case = test_helper.TestCase(
+    name='deploy_kubeflow_gcp', test_func=deploy_kubeflow_gcp)
+  test_suite = test_helper.init(
+    name='deploy_kubeflow_gcp', test_cases=[test_case])
+  test_suite.run()
+
+if __name__ == "__main__":
+  logging.basicConfig(
```
Comment (ankushagarwal) [inline testing/deploy_kubeflow_gcp.py]: test_helper.init() takes care of logging initialization

https://github.com/kubeflow/testing/blob/master/py/kubeflow/testing/test_helper.py#L159

File: testing/output/artifacts/logs/deploy_kubeflow_gcp.log [hunk 2]
```diff
@@ -0,0 +1,4 @@
+INFO|2018-05-23T18:14:50|deploy_kubeflow_gcp.py:118| Creating deployment jlewi-kubeflow-test3 in project cloud-ml-dev
```
Comment (ankushagarwal) [inline testing/output/artifacts/logs/deploy_kubeflow_gcp.log]: Checked in by accident?

**Comment by jlewi** [comment]
Most recent failure was a flake in gke_e2e trying to contact github.com

**Comment by jlewi** [comment]
Test flake  looks unrelated

```
Step 1/26 : FROM golang:1.8.2 as builder
Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

So I think this is ready for review.

**Comment by jlewi** [comment]
More random test failures

```
Traceback (most recent call last):
 File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
   "__main__", fname, loader, pkg_name)
 File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
   exec code in run_globals
 File "/mnt/test-data-volume/kubeflow-presubmit-kubeflow-e2e-gke-866-3fa104e-1651-5d37/src/kubeflow/kubeflow/testing/test_deploy.py", line 679, in <module>
   main()
 File "/mnt/test-data-volume/kubeflow-presubmit-kubeflow-e2e-gke-866-3fa104e-1651-5d37/src/kubeflow/kubeflow/testing/test_deploy.py", line 669, in main
   wrap_test(args)
 File "/mnt/test-data-volume/kubeflow-presubmit-kubeflow-e2e-gke-866-3fa104e-1651-5d37/src/kubeflow/kubeflow/testing/test_deploy.py", line 265, in wrap_test
   test_util.wrap_test(run, test_case)
 File "/mnt/test-data-volume/kubeflow-presubmit-kubeflow-e2e-gke-866-3fa104e-1651-5d37/src/kubeflow/testing/py/kubeflow/testing/test_util.py", line 88, in wrap_test
   test_case.failure = "Test failed; " + e.message
TypeError: cannot concatenate 'str' and 'ConnectionError' objects

```

**Pull request merged**